In [2]:
install.packages("ISLR")
#library(ISLR2)
#set.seed(1)
library(boot)
library(xtable)


  There is a binary version available but the source version is later:
     binary source needs_compilation
ISLR    1.2    1.4             FALSE



installing the source package 'ISLR'



We use the data of Pennsylvania re-employment bonus experiment. We are going to subset this data for tg== 4 and tg==0.

In [3]:
## loading the data
Penn <- as.data.frame(read.table("../data/penn_jae.dat", header=T ))
n <- dim(Penn)[1]
p_1 <- dim(Penn)[2]
Penn<- subset(Penn, tg== 4| tg==0)
attach(Penn)

Then T4 is defined

In [4]:
T4<- (tg==4)
summary(T4)

   Mode   FALSE    TRUE 
logical    3354    1745 

In [5]:
head(Penn)

abdt  tg inuidur1 inuidur2 female black hispanic othrace dep q1 ... q5 q6
1  10824 0  18       18       0      0     0        0       2   0  ... 1  0 
4  10824 0   1        1       0      0     0        0       0   0  ... 1  0 
5  10747 0  27       27       0      0     0        0       0   0  ... 0  0 
12 10607 4   9        9       0      0     0        0       0   0  ... 0  0 
13 10831 0  27       27       0      0     0        0       1   0  ... 1  0 
14 10845 0  27       27       1      0     0        0       0   0  ... 1  0 
   recall agelt35 agegt54 durable nondurable lusd husd muld
1  0      0       0       0       0          0    1    0   
4  0      0       0       0       0          1    0    0   
5  0      0       0       0       0          1    0    0   
12 0      1       0       0       0          0    0    1   
13 0      0       1       1       0          1    0    0   
14 0      0       1       0       0          1    0    0

## The Bootstrap

In [6]:
#write.csv(Penn,"../data/penn_jae.dat", row.names = FALSE)

In [7]:
ncol(Penn)

[1] 23

In [8]:
dim(Penn)

[1] 5099   23

In [9]:
boot.fn <- function(data, index)
  coef(lm(log(inuidur1)~T4+ (female+black+othrace+factor(dep)+q2+q3+q4+q5+q6+agelt35+agegt54+durable+lusd+husd), data = data, subset = index))

In [10]:
lm(log(inuidur1)~T4+ (female+black+othrace+factor(dep)+q2+q3+q4+q5+q6+agelt35+agegt54+durable+lusd+husd), data = Penn)


Call:
lm(formula = log(inuidur1) ~ T4 + (female + black + othrace + 
    factor(dep) + q2 + q3 + q4 + q5 + q6 + agelt35 + agegt54 + 
    durable + lusd + husd), data = Penn)

Coefficients:
 (Intercept)        T4TRUE        female         black       othrace  
    2.178462     -0.071692      0.126368     -0.293768     -0.472445  
factor(dep)1  factor(dep)2            q2            q3            q4  
    0.029867      0.096187      0.073678     -0.038507     -0.054949  
          q5            q6       agelt35       agegt54       durable  
   -0.144178      0.003361     -0.162772      0.229667      0.126557  
        lusd          husd  
   -0.175353     -0.105225  


In [11]:
boot.fn(Penn, 1:5099)

(Intercept)       T4TRUE       female        black      othrace factor(dep)1 
 2.178462326 -0.071692484  0.126368328 -0.293767980 -0.472445058  0.029866899 
factor(dep)2           q2           q3           q4           q5           q6 
 0.096186517  0.073678072 -0.038506537 -0.054949195 -0.144177912  0.003361318 
     agelt35      agegt54      durable         lusd         husd 
-0.162772168  0.229666708  0.126557359 -0.175352572 -0.105224727

 The `boot.fn()` function can also be used in order to create
 bootstrap estimates for the intercept and slope terms by randomly sampling from among the observations with replacement. Here we give two examples.

In [12]:
sample(5099, 5099, replace = T)

[1] 3372 1822 2445 3406 2026 4624 1727 4373 1403 2781 3893  323  260  329
  [15]  619  985 1652  498 2790 3854 3224 4068 1686 1484 3672 1107  208 4131
  [29] 4062  533 1871 4632   12 2144 1799 2841  214 4591 2326 1157  537  222
  [43] 1359 3323 4546 3549  142  699 2615 1885 4654 4885  270 3115 2351 2461
  [57] 2122 1944  116  946 2482  259 4587 3839 1780 3527 3885 1066 3354 1125
  [71] 4287 1445  579 4337 2839  531 4195 1037 2551  376 4542 3502 3345 3686
  [85] 3091 4716 1337 1368 3927 2479 2584 1921 2574 2176  571 4104 3130 4429
  [99] 3136 3852 2595 3923  588 2612 2426 4938 4677 4499 2204 1280 3298 3142
 [113] 3559   34 2149 2825 2870 3432 3964 4565 3639 3810 2846 1295 1763 1002
 [127] 1031 3267 1955 3508 3639 4147 1401 4849 4251 3278 1167 2731 1689  596
 [141] 1191  837 2322 2701 2850  612 3135 4087 2630 4046 1761 4129 3440 2114
 [155] 2295  636  553 4962 1075 2147 3354  213 4446  483 1487  889 4626  362
 [169]  660  358 2240 1794  464 2744  737 3662 3679 4998 1386  147  326 4221
 [183] 3693 4166 1201 1533 3063  426 3992 3481 1763 1860 4061 2544 2025 2159
 [197]  796 5084 4463 3638 3913 1190 2047 4748  297 3071 1964 2060  137 1068
 [211] 1452  257 4872 1995 2425 1146   36  633 1948 2026 4561 1629 3173 3199
 [225] 1176  526 2589 3719 3016 4355 3754 4351 2877 3441 2731 2608  649 4397
 [239] 1432 3688 2607 4090 2611  786   47    1 3086 2621 1173 4648  602  554
 [253] 3098 1803 2086 3839 1174 4674 2990 2594 3241 2016 1424 3363 4727 5023
 [267] 4580 1638 2771 1194 3406 1203 3289 3752  866 4071 4151 4087 4083  203
 [281] 2367 4603 2434 5084 4134 4343 1947 2981 4555 1131 3837 4989 3049   52
 [295] 3303 2984 1192 1982 1559 3461 4131  282 2900  753 3097 1599 3281  664
 [309]  551 3895 3380 3386  756 2776 1340 1853 1759 1940 1219 1674 2048  987
 [323] 1704 3057  776 3195 3846  496   62 4750 1009 4383 1620  812 4327  571
 [337] 1055 2561  629 3819 3901 3881 4922 1231 2024 4981  926  418  498 2675
 [351] 2243 4388 3566   75 2924 4149 1000   65 2930 4041 1329  937 3377 5076
 [365] 3854  427 3998 4692  343 3712  425 3855 2545 4979 4299 2358 1510 1119
 [379]  816  172 2151 5036 2733 4207 4909 4640  294 3488  811 2392 3648  625
 [393] 1863 1256 1416 2407 4510 4667 1093 3584 3316 2190 4196  115 1820 4072
 [407] 2820 1138  650  255 2579 1757 1704 2399 2479 2552 4266 2500 4895 1640
 [421] 1419 4997 4168  617 1885 2185 4366  969 1128  458 3540 2699 1137 2726
 [435] 2322 4996 1878 1088 2465 3023 3088 1086 2230 1597 3121 4592 2971  828
 [449] 3740  881 1330 5097 3922 2941 5075 2999 2321 1222 5085 2766 2261 2435
 [463] 2191 3503  718 2864 2942 3329 1346 2347  779 2532 4425 2418   47 3681
 [477] 2593 2739 4405  886 2842 2185 2688 1368 3977  927 2184  123  756 2733
 [491]  471 2858  479 3730 2070 3653 4930 4348  721 4559 2261 1118   84 3019
 [505]    2  818 4696 3953 1793 3747  766 2651 2144 4090 2693 3235 1874 1032
 [519] 3598 5067 3859 3751 4316 1322 2889 2358  504  680  641 2070  226 3886
 [533]  975 3484 5011  260 3967  910 2529 2582 4064 4091 2080  930 4742 4961
 [547]  956    2 2479 3655  476 4640 3251 4876  316  953 1851 2941 3101 2583
 [561] 3137 3642 1670  230 4897 1978 4286  506 3899 1643 4222 1412 1119 3896
 [575] 5046 4513 3161 4305 3057  613 2450  532 3034 3685 3236  825 3744 3607
 [589] 3503 3153 1462 2447 2224  543  752 1123 3328 3688 3153 2174  896 4503
 [603]  890  419 4195 2212   95 3983 4694 5087 3553 2344 2269 4653  801 2853
 [617] 3715 1429 4937 1210  626 1128 4025 3952 3570 3107  745 2640 1459 1050
 [631] 2328 2521 2761 1710 5060 3910  619 1016 3159 2529 1075 4108 3378 2921
 [645]  265  790 2456 4305 1951 3925 3523   54  612 2395 4932  671 3680  499
 [659] 4851 4982 3135 2357 3072 4396  975 1130 4821 3084  706 4532 1352  624
 [673]  219  305 1110 4501 1878 1539  504 3628 4219 3436 4870 4836 4213 3634
 [687] 4921 3542 1407  883 3317 4041 3866 4365  767 1169 4360 3141 1588 2243
 [701] 1496 2367 2156  706 1948 2873  371 3383 3072  274 2822 2553 4222 1014
 [715] 1365 1351   16 1882 4749 1321  773 2447 1983 3248 1045 3279 4718 4833

In [13]:
set.seed(1)
boot.fn(Penn, sample(5099, 5099, replace = T))

(Intercept)       T4TRUE       female        black      othrace factor(dep)1 
 2.258670234 -0.117228952  0.142462443 -0.374373286 -0.713971846  0.009594995 
factor(dep)2           q2           q3           q4           q5           q6 
 0.090434910  0.005622350 -0.133365893 -0.114376148 -0.180682473  0.036319276 
     agelt35      agegt54      durable         lusd         husd 
-0.153838318  0.185651787  0.085436134 -0.156716042 -0.092207747

Next, we use the `boot()` function to compute the standard errors of 1,000 bootstrap estimates for the intercept and slope terms.

In [14]:
boot(Penn, boot.fn, 1000)
B <- boot(Penn, boot.fn, 1000)


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Penn, statistic = boot.fn, R = 1000)


Bootstrap Statistics :
         original        bias    std. error
t1*   2.178462326  0.0010812733  0.15469174
t2*  -0.071692484 -0.0003483223  0.03552280
t3*   0.126368328  0.0007853971  0.03518438
t4*  -0.293767980  0.0007320124  0.06028143
t5*  -0.472445058 -0.0049861202  0.24369741
t6*   0.029866899  0.0007690384  0.05555582
t7*   0.096186517 -0.0011770733  0.04531520
t8*   0.073678072 -0.0010033034  0.15235277
t9*  -0.038506537 -0.0003146298  0.15005091
t10* -0.054949195 -0.0008440703  0.15097113
t11* -0.144177912 -0.0015435867  0.15021731
t12*  0.003361318 -0.0038177923  0.16140880
t13* -0.162772168  0.0017786902  0.03816477
t14*  0.229666708  0.0002651347  0.05836239
t15*  0.126557359  0.0001236963  0.04853387
t16* -0.175352572 -0.0006751971  0.04095015
t17* -0.105224727 -0.0007444597  0.04512871

This indicates that the bootstrap estimate for ${\rm SE}(\hat{\beta}_1)$ is $0.0355$, the bootstrap estimate for ${\rm SE}(\hat{\beta}_2)$ is $0.0351$ and that the bootstrap estimate for ${\rm SE}(\hat{\beta}_3)$ is $0.0602$.
As discussed in Section 3.1.2, standard formulas can be used to compute the standard errors for the regression coefficients in a linear model. These can be obtained using the  `summary()` function.

In [15]:
A <-summary(lm(log(inuidur1)~T4+ (female+black+othrace+factor(dep)+q2+q3+q4+q5+q6+agelt35+agegt54+durable+lusd+husd), data = Penn))$coef
A

Estimate     Std. Error t value     Pr(>|t|)    
(Intercept)   2.178462326 0.15901507 13.69972271 5.631260e-42
T4TRUE       -0.071692484 0.03546326 -2.02159887 4.327012e-02
female        0.126368328 0.03482493  3.62867401 2.876799e-04
black        -0.293767980 0.05297556 -5.54534899 3.081966e-08
othrace      -0.472445058 0.19839750 -2.38130547 1.728801e-02
factor(dep)1  0.029866899 0.05414025  0.55165799 5.812070e-01
factor(dep)2  0.096186517 0.04686228  2.05253621 4.016868e-02
q2            0.073678072 0.15682593  0.46980799 6.385124e-01
q3           -0.038506537 0.15647802 -0.24608272 8.056281e-01
q4           -0.054949195 0.15656019 -0.35097809 7.256193e-01
q5           -0.144177912 0.15588797 -0.92488160 3.550714e-01
q6            0.003361318 0.16645573  0.02019347 9.838898e-01
agelt35      -0.162772168 0.03696505 -4.40340780 1.087509e-05
agegt54       0.229666708 0.05917173  3.88135887 1.051906e-04
durable       0.126557359 0.04814585  2.62862426 8.598901e-03
lusd         -0.175352572 0.04098250 -4.27871860 1.914271e-05
husd         -0.105224727 0.04490864 -2.34308443 1.916341e-02

The standard error estimates for $\hat{\beta}_1$, $\hat{\beta}_2$ and
$\hat{\beta}_3$ obtained using the formulas from
Section 3.1.2 are $0.0354$, $0.0348$ and $0.0529$ respectively.
Interestingly, these are somewhat different from the
estimates obtained using the bootstrap.  Does this indicate a problem
with the bootstrap? In fact, it suggests the opposite.  Recall that
the standard formulas given in Equation 3.8 on page66 rely on certain assumptions. For example, they depend
on the unknown parameter $\sigma^2$, the noise variance. We then estimate $\sigma^2$
using the RSS. Now although the formulas for the standard errors do not rely on the linear model
being correct, the estimate for $\sigma^2$ does.
We see in
Figure 3.8 on page91 that there is a non-linear relationship in
the data, and so the residuals from a linear fit will be inflated, and so will $\hat{\sigma}^2$.
Secondly, the standard formulas assume (somewhat unrealistically) that the $x_i$ are fixed, and all the variability comes from the variation in the errors $\epsilon_i$.
 The bootstrap approach does not rely on any of these assumptions, and so it is
likely giving a more accurate estimate of the standard errors of
$\hat{\beta}_1$, $\hat{\beta}_2$ and $\hat{\beta}_3$ than is the `summary()`
function.

Below we compute the bootstrap standard error estimates and the standard
linear regression estimates that result from fitting the quadratic model to the data. Since this model provides a good fit to the data (Figure 3.8), there is now a better correspondence between the bootstrap estimates and the standard estimates of ${\rm SE}(\hat{\beta}_1)$, ${\rm SE}(\hat{\beta}_2)$ and ${\rm SE}(\hat{\beta}_3)$.

In [16]:
# coefcientes y errores estandar para la regresión
t4_coef <-A[2,1]
female_coef <-A[3,1]
black_coef <-A[4,1]

t4_se <-A[2,2]
female_se <-A[3,2]
black_se <-A[4,2]


In [17]:
#Cuadro de resultados

library(xtable)
table<- matrix(0, 6, 2)
table[1,1]<-   t4_coef[1]
table[1,2]<-   t4_se[1]

table[2,1]<-  female_coef[1]
table[2,2]<-  female_se[1]

table[3,1]<-   black_coef[1]
table[3,2]<-   black_se[1]

table[4,1]<-  -0.071692484
table[4,2]<-   0.03552280

table[5,1]<-   0.126368328
table[5,2]<-   0.03518438

table[6,1]<-  -0.293767980
table[6,2]<-   0.06028143





colnames(table)<- c("Coef","Standar error")
rownames(table)<- c("T4", "FEMALE", "BLACK","T4 BOOT", "FEMALE BOOT", "BLACK BOOT")
tab<- xtable(table, digits=5)
tab

Coef        Standar error
T4          -0.07169248 0.03546326   
FEMALE       0.12636833 0.03482493   
BLACK       -0.29376798 0.05297556   
T4 BOOT     -0.07169248 0.03552280   
FEMALE BOOT  0.12636833 0.03518438   
BLACK BOOT  -0.29376798 0.06028143